docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+* **Machine**: a declarative specification for an infrastructure component that hosts a Kubernetes Node. Commonly, a Machine represents a VM instance.
+* **MachinePool**: a spec for an infrastructure component that can scale a group of Machines. A MachinePool acts similarly to a [MachineDeployment][], but leverages scaling up resources that are specific to an infrastructure provider. However, while individual Machines in a MachineDeployment are visible and can be removed, individual Machines in a MachinePool are not represented.
+* **MachinePool Machine**: a Machine that is owned by a MachinePool. It requires no special treatment in Cluster API in general; a Machine is a Machine without regard to ownership.
+
+See also the [Cluster API Book Glossary][].
+
+## Summary
+
+MachinePools should be enhanced to own Machines that represent each of its `Replicas`.
+
+These "MachinePool Machines" will open up the following opportunities:
+
+- Share common behavior for Machines in MachinePools, which otherwise would need to be implemented by each infrastructure provider.
+- Allow both Cluster API-driven and externally driven interactions with individual Machines in a MachinePool.
+- Enable MachinePools to use deployment strategies similarly to MachineDeployments.
+
+## Motivation
+
+MachineDeployments and Machines provide solutions for common scenarios such as deployment strategy, machine health checks, and retired node deletion. Exposing MachinePool Machines at the CAPI level will enable similar integrations with external projects. 
+
+For example, Cluster Autoscaler needs to delete individual Machines. Without a representation in the CAPI API of individual Machines in MachinePools, projects building integrations with Cluster API would need to be specialized to delete the cloud-specific resources for each MachinePool infrastructure provider. By exposing MachinePool Machines in CAPI, Cluster Autoscaler can treat Machines built by MachineDeployments and MachinePools in a uniform fashion through CAPI APIs.
+
+As another example, currently each MachinePool infrastructure provider must implement their own node drain functionality. When MachinePool Machines become available in the CAPI API, all providers can use its existing node drain implementation.
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines.
+- Extend CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments
+- Ensure that Machines in MachinePools participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+
+### Non-Goals/Future Work
+
+- Require infrastructure providers to implement MachinePool Machines.
```

Comment:
Yes. There may be cloud providers whose native resource behind a MachinePool doesn't allow for inspecting or deleting individual instances. Also, providers won't implement MachinePool Machine support immediately. In either case, MachinePools should continue to work exactly as they do today.

Reply:
I suppose we could create Machines to represent each replica in a MachinePool, even if the provider doesn't support what's in this proposal. That would be a more consistent UX between providers, and also in comparison to MachineDeployments.

But since deleting one of those MachinePool Machines wouldn't have the effect of removing a provider-specific Machine, it would be confusing and couldn't be managed by cluster-autoscaler (which is fine--just don't annotate the MachinePool and CA will ignore it). Maybe we could prevent MachinePool Machines from being deleted in that case?

Reply:
>There may be cloud providers whose native resource behind a MachinePool doesn't allow for inspecting or deleting individual instances

Can we please add a note for that here with specific examples, so it's clear why this is proposing to support both?

Reply:
So it’s been suggested that having MPMs be optional creates a split UX, one where providers who support it have MachinePools backed by Machines, and where those don’t (yet?) have MachinePools with the current “black box” behavior. I think this is a problem with two solutions:

- Require MachinePool providers to support MPMs (eventually, according to some schedule)
- Create “hollow” Machines for each MachinePool instance/replica that don’t affect the provider resources when deleted

I’m tending toward option #1, but would love more feedback.

Reply:
I'm +1 to option 1 given that I consider this as the next stage of the experiment, but it will be great to have feedback from other MP implementers 
@richardcase @sedefsavas opinions?

Reply:
In CAPA we have 2 types of machine pool:

- `AWSMachinePool` which maps to an auto-scale group (ASG). For this type of pool we can use the [API](https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#AutoScaling.DescribeAutoScalingInstances) to get details of the individual instances/machines. We can also use termination polices to ensure on scale-in that the machine marked with the delete label is the one that is deleted. So we could go with option 1 for this.
- `AWSManagedMachinePool` which maps to an EKS node group (which in turn maps to an ASG that uses managed EC2 instances). This is like a "machine less" machine pool as you aren't meant to care about/iteract with the underlying AWS resource. You can't influence which instance is deleted on scale-in.....when you scale the node group EKS handles the instance selection/draining/deleting. However, we can get the instance/machine details (using the tags on the ASG and then the DescribeAutoScalingInstances) and so we could still create representations of the machines if needed....we just couldn't honour the delete selection. So this machine pool flavour we would need to support "machine less" or option 2?